cpp/src/arrow/compute/kernels/aggregate_basic.cc

```diff
@@ -121,6 +122,101 @@ Result<std::unique_ptr<KernelState>> CountInit(KernelContext*,
       static_cast<const CountOptions&>(*args.options));
 }
 
+// ----------------------------------------------------------------------
+// Distinct Count implementation
+
+template <typename Type>
+struct CountDistinctImpl : public ScalarAggregator {
+  using MemoTable = typename arrow::internal::HashTraits<Type>::MemoTableType;
+
+  Status consume_memo(const ArrayData& arr) {
+    auto visit_null = [&]() {
+      //return string_builder.AppendNull();
+      return Status::OK();
+    };
+    auto visit_value = [&](typename Type::c_type arg) {
+      //ARROW_ASSIGN_OR_RAISE(auto formatted, formatter(arg));
+      //return string_builder.Append(std::move(formatted));
+      int y;
+      RETURN_NOT_OK(memo_table_->GetOrInsert(arg, &y));
+      //std::cout << "AAAAAAAAAAAAAADFFFFFFFFFFFFFFFFFF\n" << arg << "\n";
+      return Status::OK();
+    };
+    RETURN_NOT_OK(VisitArrayDataInline<Type>(arr, visit_value, visit_null));
+    return Status::OK();
+  }
+
+  explicit CountDistinctImpl(CountOptions options) : options(std::move(options)), memo_table_(new MemoTable(default_memory_pool(), 0)) {
```

Comment:
Hmm, we shouldn't use the default memory pool. CountDistinctInit gets a KernelContext which can be passed here; it contains the memory pool to use.